File: colour charge visualization/colour-charge.js
var redQuarks = 0; var antiRedQuarks = 0; var redTotal = redQuarks+antiGreenQuarks+antiBlueQuarks;
var greenQuarks = 0; var antiGreenQuarks = 0; var greenTotal = greenQuarks+antiRedQuarks+antiBlueQuarks;
var blueQuarks = 0; var antiBlueQuarks = 0; var blueTotal = blueQuarks+antiRedQuarks+antiGreenQuarks;
var maximum = 0;
var waitFor = '';
var numKeys = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];

function scaleValues(value, old_max, old_min, new_max, new_min, inverse=false){
    if(maximum == 0){return 0}
    //if (inverse){return round((new_min + (new_max - new_min) * ((value - old_min) / (old_max - old_min))))}
    else{return round(255-(new_min + (new_max - new_min) * ((value - old_min) / (old_max - old_min))))}
    // effectively the map() function in other languages

}

function setup(){
    createCanvas(windowWidth, windowHeight)
}

function draw(){
    print(redQuarks, antiGreenQuarks, antiBlueQuarks, redTotal)
    redTotal = Number(redQuarks)+Number(antiGreenQuarks)+Number(antiBlueQuarks)
    greenTotal = Number(greenQuarks)+Number(antiRedQuarks)+Number(antiBlueQuarks)
    blueTotal = Number(blueQuarks)+Number(antiRedQuarks)+Number(antiGreenQuarks)

    maximum = max(max(redTotal, greenTotal), blueTotal)

    colour = [scaleValues(redTotal, maximum, 0, 0, 255), scaleValues(greenTotal, maximum, 0, 0, 255), scaleValues(blueTotal, maximum, 0, 0, 255)]
    //print('red: '+redTotal+'/'+maximum, 'green: '+greenTotal+'/'+maximum, 'blue: '+blueTotal+'/'+maximum, )
    background(colour[0], colour[1], colour[2])
    textSize(40)
    stroke(0)
    strokeWeight(3)

    xOffset = -11
    yOffset = 13

    fill(255, 0, 0); 
    circle(1*windowWidth/4, 1*windowHeight/6, 80)
    fill(0);
    text(redQuarks.toString(), (1*windowWidth/4)+xOffset, (1*windowHeight/6)+yOffset)

    fill(0, 255, 0)
    circle(2*windowWidth/4, 1*windowHeight/6, 80)
    fill(0);
    text(greenQuarks.toString(), (2*windowWidth/4)+xOffset, (1*windowHeight/6)+yOffset)

    fill(0, 0, 255)
    circle(3*windowWidth/4, 1*windowHeight/6, 80)
    fill(0);
    text(blueQuarks.toString(), (3*windowWidth/4)+xOffset, (1*windowHeight/6)+yOffset)

    fill(0, 255, 255)
    circle(1*windowWidth/4, 5*windowHeight/6, 80)
    fill(0);
    text(antiRedQuarks.toString(), (1*windowWidth/4)+xOffset, (5*windowHeight/6)+yOffset)

    fill(255, 0, 255)
    circle(2*windowWidth/4, 5*windowHeight/6, 80)
    fill(0);
    text(antiGreenQuarks.toString(), (2*windowWidth/4)+xOffset, (5*windowHeight/6)+yOffset)

    fill(255, 255, 0)
    circle(3*windowWidth/4, 5*windowHeight/6, 80)
    fill(0);
    text(antiBlueQuarks.toString(), (3*windowWidth/4)+xOffset, (5*windowHeight/6)+yOffset)

    strokeWeight(2)
    if(colour == [255, 255, 255].toString()){text('Colour charge is balanced!', -250+windowWidth/2, 25+windowHeight/2)} else{print(colour)}

    
}

function keyTyped() {

    for(i in numKeys) {
        if (key == numKeys[i]) {

            print(i)
            if (waitFor == 'r') {redQuarks = i}
            if (waitFor == 'g') {greenQuarks = i}
            if (waitFor == 'b') {blueQuarks = i}
            if (waitFor == 'ar') {antiRedQuarks = i}
            if (waitFor == 'ag') {antiGreenQuarks = i}
            if (waitFor == 'ab') {antiBlueQuarks = i}
            waitFor = ''
            //console.log('r:',redQuarks, 'g:',greenQuarks, 'b:',blueQuarks, 'ar:',antiRedQuarks, 'ag:',antiGreenQuarks, 'ab',antiBlueQuarks)
            break
        }
    }

    if (key === "q") {waitFor = 'r'}
    if (key === "w") {waitFor = 'g'}
    if (key === "e") {waitFor = 'b'}
    if (key === "a") {waitFor = 'ar'}
    if (key === "s") {waitFor = 'ag'}
    if (key === "d") {waitFor = 'ab'}

}

File: colour charge visualization/colour-charge.py
import pygame as pg

def scale(value, old_max, old_min, new_max, new_min, inverse=False):
    if maximum == 0: return 0
    if inverse: return round((new_min + (new_max - new_min) * ((value - old_min) / (old_max - old_min))))
    else: return round(255-(new_min + (new_max - new_min) * ((value - old_min) / (old_max - old_min))))
    # effectively the map() function in other languages


red = 0
antiRed = 0
green = 0
antiGreen = 0
blue = 0
antiBlue = 0
waitFor = ''

width=400
height=400
screen = pg.display.set_mode((width,height))
pg.init()

font = pg.font.Font(None, 36) # for drawing text

running = True
while running:
    maximum = max(max(red+antiGreen+antiBlue, green+antiRed+antiBlue), blue+antiGreen+antiRed)
    #print((scale(red, maximum, 0, 0, 255), scale(green, maximum, 0, 0, 255), scale(blue, maximum, 0, 0, 255)))
    screen.fill((scale(red+antiGreen+antiBlue, maximum, 0, 0, 255), scale(green+antiRed+antiBlue, maximum, 0, 0, 255), scale(blue+antiRed+antiGreen, maximum, 0, 0, 255)))
                        # both antigreen & antiblue "add" red            both antired & antiblue "add" green                both antired & antigreen "add" blue
                        # this results in antired = +green +blue, antigreen = +red +blue, antiblue = +red +green

    for e in pg.event.get():
        if e.type == pg.QUIT:
            running = False

        if e.type == pg.KEYDOWN:
            if waitFor != '':
                if e.key in range(48, 58): # number keys 0-9
                    if waitFor == 'r':
                        red = e.key - 48
                    if waitFor == 'g':
                        green = e.key - 48
                    if waitFor == 'b':
                        blue = e.key - 48
                    if waitFor == 'ar':
                        antiRed = e.key - 48
                    if waitFor == 'ag':
                        antiGreen = e.key - 48
                    if waitFor == 'ab':
                        antiBlue = e.key - 48
                    waitFor = ''

            # control which colour is being chosen
            # QWEASD = 113 119 101 97 115 100
            if e.key == 113:
                waitFor = 'r'
            if e.key == 119:
                waitFor = 'g'
            if e.key == 101:
                waitFor = 'b'
            if e.key == 97:
                waitFor = 'ar'
            if e.key == 115:
                waitFor = 'ag'
            if e.key == 100:
                waitFor = 'ab'
            
    if waitFor != '':
        #colour = (scale(red+antiGreen+antiBlue, maximum, 0, 0, 255, True), scale(green+antiRed+antiBlue, maximum, 0, 0, 255, True), scale(blue+antiRed+antiGreen, maximum, 0, 0, 255, True))
        if (scale(red+antiGreen+antiBlue, maximum, 0, 0, 255, True), scale(green+antiRed+antiBlue, maximum, 0, 0, 255, True), scale(blue+antiRed+antiGreen, maximum, 0, 0, 255, True)) == (0, 0, 0):
            colour = (255, 255, 255)
        else: colour = (0, 0, 0)
        if waitFor == 'r':
            pg.draw.circle(screen, colour, (40, 40), 25)
        if waitFor == 'g':
            pg.draw.circle(screen, colour, (width/2, 40), 25)
        if waitFor == 'b':
            pg.draw.circle(screen, colour, (width-40, 40), 25)
        if waitFor == 'ar':
            pg.draw.circle(screen, colour, (40, height-40), 25)
        if waitFor == 'ag':
            pg.draw.circle(screen, colour, (width/2, height-40), 25)
        if waitFor == 'ab':
            pg.draw.circle(screen, colour, (width-40, height-40), 25)

    offset = 10

    pos = (40, 40); pg.draw.circle(screen, (255, 0, 0), pos, 20)
    txt = font.render(str(red), True, (0, 0, 0)); screen.blit(txt, (pos[0]-offset+4, pos[1]-offset))

    pos = (width/2, 40); pg.draw.circle(screen, (0, 255, 0), pos, 20)
    txt = font.render(str(green), True, (0, 0, 0)); screen.blit(txt, (pos[0]-offset+4, pos[1]-offset))

    pos = (width-40, 40); pg.draw.circle(screen, (0, 0, 255), pos, 20)
    txt = font.render(str(blue), True, (0, 0, 0)); screen.blit(txt, (pos[0]-offset+4, pos[1]-offset))

    pos = (40, height-40); pg.draw.circle(screen, (0, 255, 255), pos, 20)
    txt = font.render(str(antiRed), True, (0, 0, 0)); screen.blit(txt, (pos[0]-offset+4, pos[1]-offset))

    pos = (width/2, height-40); pg.draw.circle(screen, (255, 0, 255), pos, 20)
    txt = font.render(str(antiGreen), True, (0, 0, 0)); screen.blit(txt, (pos[0]-offset+4, pos[1]-offset))

    pos = (width-40, height-40); pg.draw.circle(screen, (255, 255, 0), pos, 20)
    txt = font.render(str(antiBlue), True, (0, 0, 0)); screen.blit(txt, (pos[0]-offset+4, pos[1]-offset))


    if (scale(red+antiGreen+antiBlue, maximum, 0, 0, 255), scale(green+antiRed+antiBlue, maximum, 0, 0, 255), scale(blue+antiRed+antiGreen, maximum, 0, 0, 255)) == (255, 255, 255):
        txt = font.render('Colour charge is balanced!', True, (0, 0, 0)); screen.blit(txt, txt.get_rect(center=(width/2, height/2)))

    pg.display.flip()
            


pg.quit()